feat: support cluster checkpoint state request

panagiotisgts opened the pull request:
## Description

<!-- Describe the goal and purpose of this PR. -->

Support retrieving the checkpoint state per partition from RocksDB. This can be done either for `Checkpoints` or `Backups`. This functionality will be required when the checkpoint/backup schedulers are implemented, as during broker _handover of the service_ the next broker needs to determine when was the last successful tick. To do this, apart from the per partition state, we also return the _earliest checkpoint/backup_ from the aggregated partitions.

The API is exposed from the actuator under `GET /backupRuntime/state`, that will return the respective checkpoint and backup aggregate state. 

## Checklist

<!--- Please delete options that are not relevant. Boxes should be checked by reviewer. -->
- [ ] Enable backports when necessary (fex. [for bug fixes](https://github.com/camunda/camunda/blob/main/CONTRIBUTING.md#backporting-changes) or [for CI changes](https://github.com/camunda/camunda/wiki/CI-&-Automation#when-to-backport-ci-changes)).

## Related issues

closes #42243

panagiotisgts committed:
feat: support cluster checkpoint state request

Required for checkpoint/backup scheduler ticks to verify correctness and avoid possible frequent backups when SWIM fails to update cluster membership.

panagiotisgts committed:
refactor: align tests with query endpoint

panagiotisgts committed:
fix: backup request handler transition step

panagiotisgts committed:
refactor: return checkpoint state for all partitions

Alongside, the per partition status, return the earliest backup/checkpoint id from the aggregation.

Returning the earliest one allows us to be more strict about the scheduler's ticks

lenaschoenburg commented on dist/src/main/java/io/camunda/zeebe/shared/management/BackupEndpoint.java:
```diff
@@ -109,6 +114,17 @@ public WebEndpointResponse<?> delete(@Selector @NonNull final long id) {
+  private WebEndpointResponse<?> state(final String type) {
+    try {
+      final var cpType =
+          BackupApi.BACKUP.equals(type) ? CheckpointType.SCHEDULED_BACKUP : CheckpointType.MARKER;
```
🔧 I think it's surprising to users that they ask for the latest "backup" but then only see the latest scheduled backup.
I'd suggest that the actuator takes slightly different parameters: 
1. `latestBackup` -> returns the latest backup, manual or scheduled
2. `latestCheckpoint` -> returns the latest marker
And optionally we could have `latestManualBackup` and `latestScheduledBackup`, if we actually need it. For scheduling purposes, I think it doesn't matter whether the latest backup was a manual or scheduled one.


But also see my other comment, if we go with a parameter-less endpoint, we don't need to change this.

lenaschoenburg replied:
Yeah that makes sense :+1: I guess it's just a bit confusing to see `SCHEDULED_BACKUP` used rather arbitrarily here to mean "actually, any backup type".

lenaschoenburg commented on dist/src/main/java/io/camunda/zeebe/shared/management/BackupEndpoint.java:
```diff
@@ -81,19 +82,23 @@ public WebEndpointResponse<?> listAll() {
-  public WebEndpointResponse<?> query(@Selector final String prefixOrId) {
-    if (prefixOrId.endsWith(BackupApi.WILDCARD)) {
-      return listPrefix(prefixOrId);
+  public WebEndpointResponse<?> query(@Selector final String prefixOrIdOrType) {
```
🔧 I think I'd prefer a query endpoint without a parameter that returns both latest backup and latest checkpoint info. I'll use the same parameter-less endpoint to return the backup range information.

panagiotisgts replied:
Yeah maybe that makes more sense and it's much cleaner in regards to the API definition which is messy right now.

panagiotisgts committed:
fix: include both checkpoint and backup state from runtime when querying

panagiotisgts committed:
fix: extract state query to its own path

Use of selector match all to manually parse arguments allowing better control over the endpoint args

panagiotisgts committed:
docs: fix backup api definitoin

panagiotisgts committed:
fix: rebase

panagiotisgts committed:
fix: exclude missing partition state in response

lenaschoenburg commented on zeebe/protocol-impl/src/main/java/io/camunda/zeebe/protocol/impl/encoding/CheckpointStateResponse.java:
```diff
@@ -0,0 +1,142 @@
+      bodyEncoder.checkpointStatesCount(0);
+    } else {
+      final var checkpointStateEncoder = bodyEncoder.checkpointStatesCount(checkpointStates.size());
+      for (final PartitionCheckpointState partitionState : checkpointStates) {
```
🙃 the `if-else` is unnecessary, the else block works for both cases.

lenaschoenburg reviewed (approved):
Nicely done 🚀

panagiotisgts committed:
refactor: return empty instead of error state

panagiotisgts committed:
fix: return error when more arguments provided

panagiotisgts committed:
refactor: replace future handle with thenApply
